stdlib/public/core/Integers.swift.gyb

```diff
@@ -3391,6 +3391,7 @@ ${assignmentOperatorComment(x.operator, True)}
 ${overflowOperationComment(x.operator)}
   @inlinable // FIXME(sil-serialize-all)
   @_transparent
+  @_semantics("ignore.overflow")
```

Comment:
Why does this need to be transparent?

Reply:
Let me elaborate. Going to the AST level here seems like a hack. Is there any reason that we couldn't just use always inline?

Reply:
`+=`, `&+`, and `unsafeAdding` are all defined in terms of `addingReportingOverflow`, and likewise for other operators. One would continue to want this to be inlined even in -Onone, I’d imagine.

Reply:
@xwu we can just fix always inline to mean always inline.

Reply:
Last I heard, `@inline(__always)` doesn’t work with generics, is run only in -O, and is a non-mandatory hint to the performance inliner. If that’s no longer the case then it’d be wonderful, but I think these differences are what made it unsuitable for these basic operations.

I recall that a single missing `@_transparent` (for `&&`, I think) in a prerelease version of Swift slowed one of my methods by a factor of 100, so tweaks to these annotations now cause a bit of heartburn ;)

Reply:
@xwu  "in a prerelease version of Swift slowed one of my methods by a factor of 100" - that is interesting. Thanks for the input. Can you share that method if possible? It could be a good addition to the performance tests (at least may serve to document what we lost in performance). 

From the SIL diagnostic perspective, @_transparent is better since constant propagation can propagate through them to detect errors statically. Constant folding, sadly, is already quite limited due to many Integer protocol initializers not being transparent. My motivation for going along this direction is not to make  it even weaker by changing more of @_transparent to @inline. 

Looking into the AST could have been avoided if only the Inliner kept track of the inlined SIL function  in DebugLoc for mandatory inlining, as it does  for performance inlining. Unfortunately, it is not done and probably because we do not want to step into @_transparent functions during debugging. 

To summarize, as far as I can see, the following are the choices we have here: 

(a) The current AST hack:
          Pros: Seem to work. 
          Cons: 
                   - Needs AST look up in constant folding which is a hack. 
                   - Yet another special case in already complicated constant folding the has many  hacks.

(b) Use @inline(__always) instead of @_transparent: 

     Pros: We avoid adding one more hack  in constant folding
     Cons: 
           1. We may have a performance hit. We can try performance suites but would that be enough?
           2. constant folding make become weaker i.e, will give up on more cases. It may not matter, but it seems difficult to predict. 

(c) Keep track of Mandatorily inlined function information as we do for performance inlining 

     The implication of this is not clear to me and the changes will be somewhat non-local. This however has the potential to make constant folding better and eliminate some false positives with mandatorily inlined functions.

@devincoughlin Devin, can you also share your views on this?

Reply:
Some more thoughts:

My reason for pushing back strongly on this hack is that:

1. _transparent has always been a hack itself. It was meant originally to cheat on diagnostics. We really shouldn't be relying on it and I would argue that the fact it exists is a historical bug. transparent doesn't just mean inline always. It also has implications around debug info and diagnostics.
2. Once this change goes into the codebase, it is not going to be removed for a long time. I guarantee it and have seen it happen. There has to be a really good reason to insert such a hack. But if there is an obvious, easy solution, that eliminates the need of the hack, we should /not/ put in the hack.
3. Saying that constant propagation would become weaker without providing any examples is FUD. Next time please provide an example to complement that sort of argument (even a strawman would be fine = )). That being said, even with such an example, we can teach constant propagation how to recognize this semantic function and treat it appropriately. We already do this throughout the compiler.

Reply:
After some thought, rather than doing inline always, why can't you just teach constant propagation about the semantic function?

Reply:
You also missed another problem with the AST approach. We are generally trying to wean the optimizer off of directly accessing and reasoning about the AST in this way unless it is absolutely necessary. For instance, consider the changes that we have spoken about with DI where we could get rid of that direct AST introspection and just have it based off of the instructions.